### DISCUSSION TITLE
fx.Populate and provided constructors: pointers vs structs

### DISCUSSION THREAD

**Discussion by debedb** [uber-go/fx#1120] [Q&A]
I have had much grief figuring out my problem and I realized that `fx.Populate()` wants a provided constructor returning a struct, not a pointer. Why is that? Consider [this example](https://go.dev/play/p/hBSteuxBNIZ)

In this case, what is printed is:

```
{name2}
{name2}

```

If we of course, [reverse this](https://go.dev/play/p/qpdOwmljDhf), having `NewState2()` return a pointer and instead `NewState1()` return the struct, we'll end up with

```
{name1}
{name1}

```

Why is that? `fx.Populate()`'s [documentation](https://pkg.go.dev/go.uber.org/fx#Populate) is not clear on this.

**Comment by JacobOaks** [comment]
`fx.Populate`'s [documentation](https://pkg.go.dev/go.uber.org/fx#Populate) states "All targets must be pointers to the values that must be populated". So this means that if you pass `fx.Populate` a value of type `*State`, it will populate that value with a `State`. The example you linked is passing the same type to `fx.Populate` twice, of which there is only a single value in the container: that returned by `NewState2`, so both pointers you pass in will be populated with the same value.

`NewState1` returns a different type: `*State`. If you want to retrieve that, you'd need to pass a double-pointer to `fx.Populate`, a var of type `**State`:
```
var (
	state1 *State
	state2 State
)

app := fx.New(
	fx.Populate(&state1), // gets set to "name1"
	fx.Populate(&state2), // gets set to "name2"
	// ...
)
```
(see your modified example [here](https://go.dev/play/p/PCEX2kx0JAs))

If you'd rather both `NewState1` and `NewState2` return just a `State` (no pointer), I'd suggest you look into [Annotations](https://pkg.go.dev/go.uber.org/fx#Annotate), specifically [fx.ParamTags](https://pkg.go.dev/go.uber.org/fx#ParamTags) and [fx.ResultTags](https://pkg.go.dev/go.uber.org/fx#ResultTags), which allow multiple values of the same type to exist in the container. `fx.Populate` is compatible with annotations as well (see example in [fx.Populate](https://pkg.go.dev/go.uber.org/fx#Populate) documentation).
